File: 24.swap-nodes-in-pairs.py
#
# @lc app=leetcode id=24 lang=python3
#
# [24] Swap Nodes in Pairs
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        
        dummy = ListNode()
        dummy.next = head
        prev_node = dummy
        cur_node = head

        while cur_node and cur_node.next:
            
            prev_node.next = cur_node.next
            cur_node.next = prev_node.next.next
            prev_node.next.next = cur_node

            prev_node = cur_node
            cur_node = cur_node.next
        
        return dummy.next
            
            
        
# @lc code=end

File: 24.swap-nodes-in-pairs.java
/*
 * @lc app=leetcode id=24 lang=java
 *
 * [24] Swap Nodes in Pairs
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {

        if (head == null || head.next == null) {
            return head;
        }

        ListNode dummy = new ListNode();

        ListNode prevNode = dummy;
        ListNode currNode = head;

        while (currNode != null && currNode.next != null) {
            prevNode.next = currNode.next;
            currNode.next = prevNode.next.next;
            prevNode.next.next = currNode;

            prevNode = currNode;
            currNode = currNode.next;
        }

        return dummy.next;
        
    }
}
// @lc code=end

